Coding Style guide (#448)
lhr-solar/BPS: Docs/source/How_To/coding-style.rst

*************
Coding Style
*************

.. note:: 

    | CamelCase - FirstWordSecondWord
    | camelCase - firstWordSecondWord

Macros and Constants
====================

> Should be written in snakecase with all caps. If a macro utilizes arithmetic, it should be covered in parentheses to prevent compiler

Reviewer: ```suggestion
Macros and constants should be written in snakecase with all caps. If a macro utilizes arithmetic, it should be covered in parentheses to prevent the compiler
```

from assuming priority. Ex

.. code-block:: c

    #define ALL_CAPS_WITH_UNDERSCORE
    #define MATH (4 + x)
    const uint8_t FOUR_KG = 4;
    #ifdef TASKS_H //No leading underscores

Enums
=====

> Enum name should be in CamelCase and all element in enum should be all caps. They should be located in the header file. Ex:

Reviewer: ```suggestion
Enum names should be in CamelCase and all elements in enums should be in all caps. They should be located in the header file. Ex:
```

> Enum names should be in CamelCase and all elements in enums should be in all caps. They should be located in the header file. Ex:

Author: I guess this makes sense, but should there be some restrictions to using them in c files (I can't think of any right now)? Normally we use #defines inside the c file so I thought that would be enough.

.. code-block:: c

    enum ExampleEnum {
        ELEMENT_ONE;
        ELEMENT_TWO;
        ELEMENT_THREE;
        };

Structs
=======

All struct typedef names should have a descriptive name written in CamelCase and an appended ``_t``. The name of the struct should be

> in CamelCase and it's members should be written in camelCase to make it easier to understand gcc error messages. 

Reviewer: ```suggestion
in CamelCase and its members should be written in camelCase to make it easier to understand gcc error messages. 
```

> They should be located in the header file. Ex:

Author: Same for this, not sure why we would define a struct only for use in the c file.

.. code-block:: c

    typedef struct StructName {
        uint8_t element;
        uint16_t elementTwo;
    } StructName_t;

Unions
======

Unions should only be used when absolutely necessary (normally for communication protocols). They should be located in the
header file, written in CamelCase, and appended with an ``_t``. Members of the union should be in camelCase.

.. code-block:: c

    typedef union {
	    uint8_t b;
	    uint16_t h;
	    uint32_t w;
	    float f;
    } CANData_t;

Libraries
=========

All libraries are separated into source files and header files. The source file consists of the source code and the header file
holds the function declarations for the library. 

=============
Library Names
=============

> The name of the library depends on the layer that library is part of. 

Reviewer: ```suggestion
The name of the library depends on the layer that the library is part of. 
```

Board Support Package (BSP)
---------------------------

BSP library file names should describe the feature of the microcontroller that is being used. They should all be in 

> snakecase with the first section being ``BSP_``. If the BSP library is for a specific functionality in the system, the name should

Reviewer: ```suggestion
snakecase with the first section being ``BSP_``. If the BSP library is for specific functionality in the system, the name should
```

describe the functionality. Ex:

.. code-block:: c

    BSP_ADC.c //Microcontroller Feature
    BSP_ADC.h
    BSP_Fans.c //Interface with system

>     BSP_Fans.h

Author: Yea, makes sense.

Drivers
-------

Driver library file names should be the name of the component the library is for with the appropriate suffix. Ex:

.. code-block:: c

    Partname.c
    Partname.h
    LTC6811.c
    LTC6811.h

Application
-----------

> Application level library names should describe the functionality of the application. If the application library is for a task,

Reviewer: ```suggestion
Application-level library names should describe the functionality of the application. If the application library is for a task,
```

> it should be prefaced with the ``Task_`` prefix. the tasks name should be in CamelCase. Ex:

Reviewer: ```suggestion
it should be prefixed with the ``Task_`` prefix. the task's name should be in CamelCase. Ex:
```

.. code-block:: c 

    EEProm.c //interface with EEPROM
    Temperature.c
    Task_TaskName.c

============
Source Files
============

Source files should be formatted in the following method.

    1. Trademark/Description of file
    2. Includes
    3. Defines
    4. Globals
    5. Static Functions
    6. Public Functions
    7. Newline

The trademark/description should adhere to the following format:

.. code-block:: c

    /* Copyright (c) <YEAR> UT Longhorn Racing Solar */
    /** DRIVER.c
    * DESCRIPTION
    */

Included files should only be what is necessary for the source code to run. Defines should follow Macro coding style.
If there is debugging functionality in the source file (that is not essential for the system to run), ``#ifdef`` and ``endif``
should surround all code necessary for that functionality. This consists of included files, ``printf`` statements, and other code
inside functions that run at regular runtime. Debugging functions do not have to be surrounded. Static functions must be declare

> before all public functions. It is the prerogative of the programmer as to have the static function definition at the start of the 

Reviewer: ```suggestion
before all public functions. It is the prerogative of the programmer to have the static function definition at the start of the 
```

> source file or at the end.

Reviewer: ```suggestion
source file or the end.
```

============
Header Files
============

Header files should be formatted in the following method.

    1. Trademark/Description of file
    2. Defines
    3. Includes
    4. Typedefs
    5. Function Declarations
    6. Endifs
    7. Newline

All header files should be surrounded with ``#ifndef`` and ``#endif``.

> Header files should only have amount of includes necessary for header file to work. For example, if a function returns an 

Reviewer: ```suggestion
Header files should only have the number of includes necessary for the header file to work. For example, if a function returns an 
```

``int``, then the header file should ``#include <stdint.h>``. However, if the source file calls ``memcpy`` in a function, it is not
necessary for the header file to ``#include <stdlib.h>``.

Functions
=========

============
Descriptions
============

All function declarations and definitions should have a comment paragraph description that follows the specified format.::

    /**
     * NOTE: Include anything important someone else will need to know
     * @brief Give Description of Function
     * @param variable Describe input parameters
     * @return What function returns
     */

================
Public Functions
================

> The first section should be the layer that the function resides in. The second section should

Reviewer: ```suggestion
The first section should be the layer that which the function resides. The second section should
```

be separated by snakecase and have the library the function is a part of. The third section should

> also be separated by snakecase and describe the functions purpose. If there are multiple words in

Reviewer: ```suggestion
also be separated by snakecase and describe the function's purpose. If there are multiple words in
```

the third section they should be written in CamelCase. Ex

.. code-block:: c

    Layer_Library_Function();
    BSP_Contactor_GetState();

Descriptions for public functions should be included above the function definition in the source
file and above the declaration in the header file.

=================
Private Functions
=================

Code duplication should be avoided by using private functions instead. Their format is the same as public functions but 
do not need to have the layer section in their name. Their description should be included in the source file, not the header file.

.. code-block:: c

    /**
     * NOTE: Include anything important someone else will need to know
     * @brief Give Description of Function
     * @param variable Describe input parameters
     * @return What function returns
     */
    static void ADC_InitDMA(void);

Variables
=========

================
Local Variables
================

If variables have the same type, declare them on the same line if possible. 

> Meaningful names for variables that aren't used for iterations in loop. Should be written in camelCase.

Reviewer: ```suggestion
Meaningful names for variables that aren't used for iterations in a loop should be written in camelCase.
```

> Unecessary variables should be avoided. i.e

Reviewer: ```suggestion
Unnecessary variables should be avoided. i.e
```

.. code-block:: c
    :linenos:

    //NO BUENO
    int x = function();
    if (x) return;
    //YES BUENO

>     if (function()) return;

Reviewer 2: You aren't following your rule for curly braces
```suggestion
    if (function()) {
        return;
    }
```

> If functionality of code is affected by this (calling same function multiple time), then

Reviewer: ```suggestion
If the functionality of the code is affected by this (calling the same function multiple times), then
```

it is fine to have a variable for the return value.

================
Global Variables
================

Variable names should be meaningful and written in CamelCase. Efforts must be made to prevent global variables from
being used in multiple libraries. All global variables must be static variables unless the logic requires global variables
being shared. All global variables should have comments describing their purpose (what they hold). Ex.

.. code-block:: c
    :linenos:

    static OS_MUTEX VoltageMutex;
    static uint16_t VoltageVal[NUM_BATTERY_MODULES]; //Voltage values gathered

Miscellaneous
=============

============
Punctuation
============

> **Curly Braces:** Opening braces should be on same line as condition or function that contains them. Curly braces should always 

Reviewer: ```suggestion
**Curly Braces:** Opening braces should be on the same line as the condition or function that contains them. Curly braces should always 
```

> be used in coditional even if it is one line

Reviewer: ```suggestion
be used in conditionals even if it is one line
```

> **Parentheses:** Should be used for clarifying order of precedence.

Reviewer: ```suggestion
**Parentheses:** Should be used for clarifying the order of precedence.
```

**Indentation:** Indentation should be used in all conditionals, iteratives, and functions. Indentation should be 4 spaces (NOT TABS).
**Line Endings:** All End Of Line sequences should be LF(line feed). This prevents files from looking like they are changed when
nothing was changed at all.

===============
Common Practice
===============

> **Pointers:** Members to pointers should be accessed through ``p->member`` operator instead of ``(p).member``.

Reviewer 2: ```suggestion
**Pointers:** Members to pointers should be accessed through ``p->member`` operator instead of ``(*p).member``.
```
I assume this was a typo

**Indentation:** Tabs should be 4 spaces. If a pull request is made and changes are made to files you did not edit, check to see

> if your editor is editing whitespace when opening files (for MAC users). If these issues are not fixed, your PR WILL NOT BE MERGED

Reviewer 2: This also happens on Windows, I think

    #ifdef TASKS_H //No leading underscores

Enums
=====

Enum names should be in CamelCase with ``_e`` postfixed and all elements in enums should be in all caps. They should be located in the 
header file unless they are only used within the library. Ex:

.. code-block:: c

    typedef enum {
        ELEMENT_ONE;
        ELEMENT_TWO;
        ELEMENT_THREE;
        }ExampleEnum_e;

Structs
=======

All struct typedef names should have a descriptive name written in CamelCase and an appended ``_t``. The name of the struct should be
in CamelCase and its members should be written in camelCase to make it easier to understand gcc error messages. 
They should be located in the header file. Ex:

.. code-block:: c

    typedef struct StructName {
        uint8_t element;
        uint16_t elementTwo;
    } StructName_t;

Unions
======

Unions should only be used when absolutely necessary (normally for communication protocols). They should be located in the
header file, written in CamelCase, and appended with an ``_t``. Members of the union should be in camelCase.

.. code-block:: c

    typedef union {
	    uint8_t b;
	    uint16_t h;
	    uint32_t w;
	    float f;
    } CANData_t;

> Necessary Case: Since we could be sending either 8 bits, or 64 bits, it is necessary to have a union that packs all the diffferent

Reviewer 2: ```suggestion
Necessary Case: Since we could be sending either 8 bits, or 64 bits, it is necessary to have a union that packs all the different
```

and using the proper variable at the proper time.

.. code-block:: c

    typedef union {
    	uint8_t b;
    	uint16_t h;
    	uint32_t w;
    	uint64_t d;
    } CANData_t;

Usage:

.. code-block:: c

    CANData_t Data;
    uint8_t byte = 7;
    Data.b = byte;

Unnecessary Case: This union was created so the values of each fault could be accessed either through the bitmap or through the struct
within the union. In one case, the bitmap could be read by checking each bit in the byte sized variable, or you could check each
individual Fault value to see if it is set to 1. This is unncecessary and makes the code more difficult to read and use. 

.. code-block:: c

    typedef union{
        uint8_t bitmap;
        struct{
            State Fault_OS : 1;         // for OS faults
            State Fault_UNREACH : 1;    // for unreachable conditions
            State Fault_TRITIUM : 1;      // for errors sent from the tritium
            State Fault_READBPS : 1;    // for unsuccessfully reading from BPS CAN
            State Fault_DISPLAY : 1;    // for display faults
        };

>     } fault_bitmap_t;

Reviewer 2: LMAO I wonder where this example came from...

and some at the end).

Variables
=========

================
Local Variables
================

If variables have the same type, declare them on the same line if possible. 
Meaningful names for variables that aren't used for iterations in a loop should be written in camelCase.
Unnecessary variables should be avoided. i.e

.. code-block:: c
    :linenos:

    //NO BUENO
    int x = function();
    if (x) return;
    //YES BUENO
    if (function()) {
        return;
    }

If the functionality of the code is affected by this (calling the same function multiple times), then
it is fine to have a variable for the return value.

================
Global Variables
================

Variable names should be meaningful and written in CamelCase. Efforts must be made to prevent global variables from
being used in multiple libraries. All global variables must be static variables unless the logic requires global variables
being shared. All global variables should have comments describing their purpose (what they hold). Ex.

.. code-block:: c
    :linenos:

    static OS_MUTEX VoltageMutex;

>     static uint16_t VoltageVal[NUM_BATTERY_MODULES]; //Voltage values gathered

Reviewer 2: I don't like this variable name though... I think `Voltages[NUM_BATTERY_MODULES]` would be better

    EEPROM.c //interface with EEPROM
    Temperature.c

> Core

Reviewer 2: Why not just call this the task layer instead of creating a new name?

Author: It is similar to the FreeRTOS/STM cube code generator so if we want to port it over it wouldn't be too much of a hassle.

Reviewer 2: I think "core" is broader and less descriptive than "task" or "tasks" though, since I think we are just going to put tasks in this layer

Included files should only be what is necessary for the source code to run. Defines should follow Macro coding style.
If there is debugging functionality in the source file (that is not essential for the system to run), ``#ifdef DEBUG`` and ``endif``
should surround all code necessary for that functionality. This consists of included files, ``printf`` statements, and other code
inside functions that run at regular runtime. Debugging functions do not have to be surrounded. Static functions must be declared
before all public functions. It is the prerogative of the programmer to have the static function definition at the start of the 

> source file or the end. More information can be found in the Private Functions Section.

Reviewer 2: Can we hyperlink to the private functions section here?
